internal/scheduler/scheduler.go

```diff
@@ -273,13 +285,39 @@ func (s *Scheduler) syncState(ctx context.Context) ([]*jobdb.Job, error) {
 		jobsToUpdateById[jobId] = job
 	}
 
+	// Ensure queued jobs being retried have node anti affinities for any nodes they have already run on
+	for _, job := range jobsToUpdateById {
+		if job.Queued() && job.NumAttempts() > 0 {
+			schedulingInfoWithNodeAntiAffinity, err := s.createSchedulingInfoWithNodeAntiAffinityForAttemptedRuns(job)
+			if err != nil {
+				log.Warnf("unable to set node anti affinity for job %s because %s", job.GetId(), err)
+			} else {
+				job = job.WithJobSchedulingInfo(schedulingInfoWithNodeAntiAffinity)
+				jobsToUpdateById[job.GetId()] = job
+			}
+		}
+	}
+
 	// Jobs with no active runs are queued.
 	for _, job := range jobsToUpdateById {
 		// Determine if the job needs to be re-queued.
 		// TODO: If have an explicit queued message, we can remove this code.
 		run := job.LatestRun()
 		desiredQueueState := false
-		requeueJob := run != nil && run.Returned() && job.NumReturned() <= s.maxLeaseReturns
+		requeueJob := run != nil && run.Returned() && job.NumAttempts() < s.maxAttemptedRuns
+		if requeueJob && !job.Queued() && run.Returned() && run.RunAttempted() {
```

Comment:
can we simplify this?  To work out if the job is going to fall into this, we've got 6 bools to check!

To tie this in with my previous comment- could we not have a single loop and do something like:
```
if job.Queued || requeueJob {
  // Ensure affinity logic is set
}
```